builtin/credential/aws/path_image_tag.go

```diff
@@ -0,0 +1,299 @@
+				Type:        framework.TypeString,
+				Description: "AMI name to create a tag for.",
+			},
+
+			"policies": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "Policies to be associated with the tag.",
+			},
+
+			"max_ttl": &framework.FieldSchema{
+				Type:        framework.TypeDurationSecond,
+				Default:     0,
+				Description: "The maximum allowed lease duration",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.UpdateOperation: b.pathImageTagUpdate,
+		},
+
+		HelpSynopsis:    pathImageTagSyn,
+		HelpDescription: pathImageTagDesc,
+	}
+}
+
+// pathImageTagUpdate is used to create an EC2 instance tag which will
+// identify the Vault resources that the instance will be authorized for.
+func (b *backend) pathImageTagUpdate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	imageID := strings.ToLower(data.Get("name").(string))
+	if imageID == "" {
+		return logical.ErrorResponse("missing image name"), nil
+	}
+
+	// Parse the given policies into a slice and add 'default' if not provided.
+	// Remove all other policies if 'root' is present.
+	policies := policyutil.ParsePolicies(data.Get("policies").(string))
+
+	// Fetch the image entry corresponding to the AMI name
+	imageEntry, err := awsImage(req.Storage, imageID)
+	if err != nil {
+		return nil, err
+	}
+	if imageEntry == nil {
+		return logical.ErrorResponse("image entry not found"), nil
+	}
+
+	// If RoleTag is empty, disallow creation of tag.
+	if imageEntry.RoleTag == "" {
+		return logical.ErrorResponse("tag creation is not enabled for this image"), nil
+	}
+
+	// Create a random nonce
+	nonce, err := createRoleTagNonce()
+	if err != nil {
+		return nil, err
+	}
+
+	// max_ttl for the role tag should be less than the max_ttl set on the image.
+	maxTTL := time.Duration(data.Get("max_ttl").(int)) * time.Second
+
+	// max_ttl on the tag should not be greater than the system view's max_ttl value.
+	if maxTTL > b.System().MaxLeaseTTL() {
+		return logical.ErrorResponse(fmt.Sprintf("Registered AMI does not have a max_ttl set. So, the given TTL of %d seconds should be less than the max_ttl set for the corresponding backend mount of %d seconds.", maxTTL/time.Second, b.System().MaxLeaseTTL()/time.Second)), nil
+	}
+
+	// If max_ttl is set for the image, check the bounds for tag's max_ttl value using that.
+	if imageEntry.MaxTTL != time.Duration(0) && maxTTL > imageEntry.MaxTTL {
+		return logical.ErrorResponse(fmt.Sprintf("Given TTL of %d seconds greater than the max_ttl set for the corresponding image of %d seconds", maxTTL/time.Second, imageEntry.MaxTTL/time.Second)), nil
+	}
+
+	if maxTTL < time.Duration(0) {
+		return logical.ErrorResponse("max_ttl cannot be negative"), nil
+	}
+
+	// Attach version, nonce, policies and maxTTL to the role tag value.
+	rTagValue, err := prepareRoleTagPlainValue(&roleTag{Version: roleTagVersion,
+		ImageID:  imageID,
+		Nonce:    nonce,
```

Comment:
It doesn't look like anything is enforcing that this is truly used just once, right? It also feels a little weird to attach to the AMI name (`imageID := strings.ToLower(data.Get("name").(string))` above) rather than the ID, because the name can be chosen by a user.

For example, with this scheme, what's to stop me from deleting the authorized AMI, saving its role tag, creating another (evil, muahahahaha :smiling_imp:) AMI with the same name (which I can choose when I create it), and sticking the role tag on the new image?